Adding Terraform Enterprise security model (#1763)
hashicorp/terraform-website: content/source/docs/enterprise/before-installing/index.html.md

~> **Important:** We do not recommend running Docker under a 2.x kernel.

### IAM Policies

If you have chosen the [external services operational mode](https://www.terraform.io/docs/enterprise/before-installing/index.html#operational-mode-decision), Terraform Enterprise will need access to an S3-compliant endpoint for object storage. You can grant access to the object storage endpoint by either assigning an AWS instance profile or an equivalent IAM system in non-AWS environments.

> #### Using the Instance Profile as Default Credentials for Terraform Runs

Reviewer: ```suggestion
#### Workspace Credentials
```

> You can use Terraform Enterprise's instance profile to provide default credentials to workspaces. Terraform will attempt to use the instance profile to provision resources when you do not set credentials as environment variables. However, this approach presents a few security risks: 

Reviewer: I know I know, I already edited this! But I reread this and think this might flow better and be cleaner. The point (as I see it) is that the user has two choices, and I want to make them both clear here. This is just me reordering the nice work you already did here!

```suggestion
To set credentials for workspaces, you can: 
- Add credentials as environment variables in each workspace. 
    ~> **Important:** We highly recommend that you [restrict build worker metadata access](../system-overview/security-model.html#restrict-terraform-build-worker-metadata-access) to prevent workspaces from accessing the instance profile.
- Use Terraform Enterprise's instance profile to provide default credentials. Terraform will attempt to use the instance profile to provision resources in all workspaces that do not have credentials as environment variables. This approach presents a few security risks: 
```

Author: Hmmm... I agree that this reads better, but now I'm wondering if this really belongs in the "before installing" documentation. TFE users don't really need to think about workspace credentials until after installation.  My goal was to simply add a disclaimer to this existing documentation ("if you do this, be aware of these security implications") rather than provide much instruction.

So I kinda prefer the previous text. What do you think? Any better way to arrange this content with the new disclaimer?

Reviewer: ahh okay yes, that's fair. So I have two thoughts. First, maybe we're trying to do too much here (as you pointed out). Maybe all we need to do is be like "hey, think through the permissions you're setting in your instance profile, because they could be used to provision resources down the line." To me, that means we cut this entire section and just have a note under the AWS instance profile sentence. 

 "~>**Important:**  Terraform uses the instance profile as default credentials for workspaces that do not have credentials set as environment variables. That means Terraform could use the permissions in the instance profile to provision resources."

But then - this nice section we made is good information. AND I can't really find it anywhere else (am I missing the page it's on? Is there a good page for us to add this to? I feel like users should know it!

Author: Ok, here's my latest (and greatest?) idea on this: I'd like to keep all this content, since I think it is valuable and not stated elsewhere in our docs. I've rearranged the sections though - the heading is "IAM Policies", and the section goes on to explain the required s3 and kms policies, plus the option to use the instance profile to provide default credentials.

How does that sound?

Reviewer: okay! Works for me - thank you :-) I just suggest shortening the title to get rid of "using" at the front (see my comment). We're trying to move away from that in general, and this way the title matches the others more closely. Otherwise, feel free to merge today!

> 1. All workspaces will have the same permissions because they have access to the same instance profile. You cannot selectively allow or deny access to the instance profile for each workspace.

Reviewer: just make sure these get indented under the block above!

Reviewer: I'm going to remove this because I can tell this is going to get nixed from this section!

2. Workspaces will share the instance profile with the Terraform Enterprise application. All workspaces within the application will have access to any resources that Terraform Enterprise depends on, such as its S3 bucket, KMS keys, etc.

> ~> **Important:** If you choose not to use the instance profile for default credentials, we highly recommend that you [restrict build worker metadata access](../system-overview/security-model.html#restrict-terraform-build-worker-metadata-access) to prevent workspaces from accessing the instance profile.

Reviewer: Removing because I added it back in above!

```suggestion
```

Reviewer: same here :-)

}
```

> #### Using the Instance Profile as Default Credentials for Terraform Runs

Reviewer: ```suggestion
#### Instance Profile as Default Credentials
```
